gcsfs/retry.py

```diff
@@ -0,0 +1,133 @@
+    """Raised when the md5 hash of the content does not match the header."""
+
+    pass
+
+
+RETRIABLE_EXCEPTIONS = (
+    requests.exceptions.ChunkedEncodingError,
+    requests.exceptions.ConnectionError,
+    requests.exceptions.ReadTimeout,
+    requests.exceptions.Timeout,
+    requests.exceptions.ProxyError,
+    requests.exceptions.SSLError,
+    requests.exceptions.ContentDecodingError,
+    google.auth.exceptions.RefreshError,
+    aiohttp.client_exceptions.ClientError,
+    ChecksumError,
+)
+
+
+def is_retriable(exception):
+    """Returns True if this exception is retriable."""
+    errs = list(range(500, 505)) + [
+        # Request Timeout
+        408,
+        # Too Many Requests
+        429,
+    ]
+    errs += [str(e) for e in errs]
+    if isinstance(exception, HttpError):
+        return exception.code in errs
+
+    return isinstance(exception, RETRIABLE_EXCEPTIONS)
+
+
+def validate_response(status, content, path, headers=None):
+    """
+    Check the requests object r, raise error if it's not ok.
+
+    Parameters
+    ----------
+    r: requests response object
+    path: associated URL path, for error messages
+    """
+    if status >= 400:
+        error = None
+        if hasattr(content, "decode"):
+            content = content.decode()
+        try:
+            error = json.loads(content)["error"]
+            msg = error["message"]
+        except json.decoder.JSONDecodeError:
+            msg = content
+
+        if status == 404:
+            raise FileNotFoundError
+        elif status == 403:
+            raise IOError("Forbidden: %s\n%s" % (path, msg))
+        elif status == 502:
+            raise requests.exceptions.ProxyError()
+        elif "invalid" in str(msg):
+            raise ValueError("Bad Request: %s\n%s" % (path, msg))
+        elif error:
+            raise HttpError(error)
+        elif status:
+            raise HttpError({"code": status, "message": msg})  # text-like
+        else:
+            raise RuntimeError(msg)
+
+
+@decorator
+async def retry_request(func, retries=6, *args, **kwargs):
+    for retry in range(retries):
+        try:
+            if retry > 0:
+                await asyncio.sleep(min(random.random() + 2 ** (retry - 1), 32))
+            return await func(*args, **kwargs)
+            break
+        except (
+            HttpError,
+            requests.exceptions.RequestException,
+            google.auth.exceptions.GoogleAuthError,
+            ChecksumError,
+            aiohttp.client_exceptions.ClientError,
+        ) as e:
+            if (
+                isinstance(e, HttpError)
+                and e.code == 400
+                and "requester pays" in e.message
+            ):
+                msg = "Bucket is requester pays. Set `requester_pays=True` when creating the GCSFileSystem."
+                raise ValueError(msg) from e
+            if retry == retries - 1:
+                logger.exception("_call out of retries on exception: %s" % e)
```

Comment:
This is no longer exclusively called from `_call`

Reply:
Good catch.

Reply:
updated to `func.__name__`